src/app/CommandHandler.cpp

```diff
@@ -65,25 +65,19 @@ CHIP_ERROR CommandHandler::AllocateBuffer()
 CHIP_ERROR CommandHandler::OnInvokeCommandRequest(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
                                                   System::PacketBufferHandle && payload)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
     System::PacketBufferHandle response;
-
     VerifyOrReturnError(mState == CommandState::Idle, CHIP_ERROR_INCORRECT_STATE);
 
     // NOTE: we already know this is an InvokeCommand Request message because we explicitly registered with the
     // Exchange Manager for unsolicited InvokeCommand Requests.
-
     mpExchangeCtx = ec;
 
-    err = ProcessInvokeRequest(std::move(payload));
-    SuccessOrExit(err);
-
-    err = SendCommandResponse();
-    SuccessOrExit(err);
+    // Use the RAII feature, if this is the only Handle when this function returns, DecrementHoldOff will trigger sending response.
+    Handle workHandle(this);
+    ReturnErrorOnFailure(ProcessInvokeRequest(std::move(payload)));
+    mpExchangeCtx->WillSendMessage();
```

Comment:
Again, if we already have pending work and then hit a decoding error, this will end up closing the exchange.  But wouldn't we expect to actually send the part of the response that corresponds to the work we actually did?

As in, WillSendMessage() seems like it should be _before_ the ProcessInvokeRequest() call.